File: day20/main.go
package main

import (
	"aocutil"
	"math"
)

func part1(input int) int {
	presentsPerElf := 10
	presents := 0
	house := int(math.Sqrt(float64(input))) - 1
	for presents < input {
		house++
		presents = 0
		max := int(math.Sqrt(float64(house)))
		for i := 1; i <= max; i++ {
			if house%i == 0 {
				//house = 100, i = 5 (add for presents for both 5 and 20)
				factor := house / i
				if factor != i {
					presents += (factor * presentsPerElf)
				}
				presents += (i * presentsPerElf)
			}
		}
	}
	return house
}

func part2(input int) int {
	presentsPerElf := 11
	presents := 0
	house := int(math.Sqrt(float64(input))) - 1
	elves := make(map[int]int, 300)
	for presents < input {
		house++
		presents = 0
		max := int(math.Sqrt(float64(house)))
		for i := 1; i <= max; i++ {
			if house%i == 0 {
				factor := house / i
				if factor != i {
					if _, ok := elves[factor]; !ok {
						elves[factor] = 0
					}
					if elves[factor] < 50 {
						elves[factor]++
						presents += (factor * presentsPerElf)
					}
				}
				if _, ok := elves[i]; !ok {
					elves[i] = 0
				}
				if elves[i] < 50 {
					elves[i]++
					presents += (i * presentsPerElf)
				}
			}
		}
	}
	return house
}

func main() {
	input := 34000000
	aocutil.AOCFinish(part1(input), part2(input))
}

File: day20/main.c
#include <stdio.h>
#define INPUT 34000000
#define MAX 1000000

static int solve(const int steps, const int multiplier) {
  int presents[MAX] = {0};
  for (int i = 1; i < MAX; i++) {
    int j = i;
    int step = 0;
    while (j < MAX && step < steps) {
      presents[j] += multiplier * i;
      j += i;
      step++;
    }
  }

  for (int i = 0; i < MAX; i++) {
    if (presents[i] >= INPUT)
      return i;
  }
  return -1;
}

int main(void) {
  printf("%d\n%d\n", solve(MAX, 10), solve(50, 11));
}
